sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala

```diff
@@ -0,0 +1,292 @@
+    // Initialize new Countings instance here.
+    Countings()
+  }
+
+  private def evalPercentages(expr: Expression): (Boolean, Seq[Number]) = {
+    val (isArrayType, values) = (expr.dataType, expr.eval()) match {
+      case (_, n: Number) => (false, Array(n))
+      case (_, d: Decimal) => (false, Array(d.toDouble.asInstanceOf[Number]))
+      case (ArrayType(baseType: NumericType, _), arrayData: ArrayData) =>
+        val numericArray = arrayData.toObjectArray(baseType)
+        (true, numericArray.map { x =>
+          baseType.numeric.toDouble(x.asInstanceOf[baseType.InternalType]).asInstanceOf[Number]
+        })
+      case other =>
+        throw new AnalysisException(s"Invalid data type ${other._1} for parameter percentage")
+    }
+
+    require(values.forall(value => value.doubleValue() >= 0.0 && value.doubleValue() <= 1.0),
+      s"Percentage values must be between 0.0 and 1.0, current values = ${values.mkString(", ")}")
+
+    (isArrayType, values)
+  }
+
+  override def update(buffer: Countings, input: InternalRow): Unit = {
+    val key = child.eval(input).asInstanceOf[Number]
+    buffer.add(key)
+  }
+
+  override def merge(buffer: Countings, other: Countings): Unit = {
+    buffer.merge(other)
+  }
+
+  override def eval(buffer: Countings): Any = {
+    generateOutput(buffer.getPercentiles(percentages))
+  }
+
+  private def generateOutput(results: Seq[Double]): Any = {
+    if (results.isEmpty) {
+      null
+    } else if (returnPercentileArray) {
+      new GenericArrayData(results)
+    } else {
+      results.head
+    }
+  }
+
+  override def serialize(obj: Countings): Array[Byte] = {
+    Percentile.serializer.serialize(obj, child.dataType)
+  }
+
+  override def deserialize(bytes: Array[Byte]): Countings = {
+    Percentile.serializer.deserialize(bytes, child.dataType)
+  }
+}
+
+object Percentile {
+  object Countings {
+    def apply(): Countings = Countings(new OpenHashMap[Number, Long])
+
+    def apply(counts: OpenHashMap[Number, Long]): Countings = new Countings(counts)
+  }
+
+  /**
+   * A class that stores the numbers and their counts, used to support [[Percentile]] function.
+   */
+  class Countings(val counts: OpenHashMap[Number, Long]) extends Serializable {
+    /**
+     * Insert a key into countings map.
+     */
+    def add(key: Number): Unit = {
+      // Null values are ignored in countings.
+      if (key != null) {
+        counts.changeValue(key, 1L, _ + 1L)
+      }
+    }
+
+    /**
+     * In place merges in another Countings.
+     */
+    def merge(other: Countings): Unit = {
+      other.counts.foreach { pair =>
+        counts.changeValue(pair._1, pair._2, _ + pair._2)
+      }
+    }
+
+    /**
+     * Get the percentile value for every percentile in `percentages`.
+     */
+    def getPercentiles(percentages: Seq[Number]): Seq[Double] = {
+      if (counts.isEmpty) {
+        return Seq.empty
+      }
+
+      val sortedCounts = counts.toSeq.sortBy(_._1)(new Ordering[Number]() {
```

Comment:
Maybe a dumb question: How can we order a sequence of `Number` using the `Ordering[NumericType#InternalType]` ?